kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/queues/SchedulingQueue.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.api.async.internals.queues;
+
+import dev.responsive.kafka.api.async.internals.events.AsyncEvent;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+
+/**
+ * A non-blocking queue for async events waiting to be passed from the StreamThread to
+ * the async thread pool and scheduled for execution. This queue is not thread safe and
+ * should be owned and exclusively accessed by the StreamThread. Events that are
+ * processable -- that is, not blocked on previously scheduled events with
+ * the same key that have not yet been fully processed -- will be polled from
+ * this queue and then "scheduled" by passing them on to the {@link ProcessingQueue}
+ * <p>
+ * Threading notes:
+ * -Should only be accessed from the StreamThread
+ * -One per physical AsyncProcessor instance
+ *   (ie per logical processor per partition per StreamThread)
+ */
+public class SchedulingQueue<KIn> {
```

Comment:
we should really have a back-pressure mechanism at this level. Otherwise what can happen is we'll enqueue a huge number of events in this queue while it slowly drains. I'm thinking something simple where we just set a cap on the max number of outstanding events and block `offer` when we hit that number.

Reply:
Yep, good call